### ISSUE TITLE
Placeholder text is harmful

### ISSUE THREAD

**Issue by garrett** [cockpit-project/cockpit-podman#288]
I saw that placeholder text is being used in place of labels in this screenshot from [the Cockpit 210 / Cockpit-Podman 12 release announcement](https://cockpit-project.org/blog/cockpit-210.html):

![podman-cpu-limit](https://user-images.githubusercontent.com/10246/72060351-167d7e00-32d4-11ea-899e-df982a819676.png)

Placeholders should only be used as an example of what to type. Labels, hints, and other miscellaneous information should exist _outside_ of the input field itself.

References:
- Nielson Norman Group on placeholders: https://www.nngroup.com/articles/form-design-placeholders/
- PatternFly 4's documentation on form usage and behavior: https://www.patternfly.org/v4/design-guidelines/usage-and-behavior/forms#provide-the-right-kind-of-help (note the "helper text" block, but read through the rest of the "Provide the right kind of help" section)

**Comment by garrett** [comment]
This is an both an accessibility and usability issue.

(It's a usability issue as the text disappears when trying to type and also, in some case, the text is also truncated.)

**Comment by garrett** [comment]
There are also additional problems with the form using very wide entries for numbers. The input should be the approximate length of the expected value (plus a little bit of padding, to ensure numbers are not truncated).

**Comment by marusak** [comment]
It shows what is the default value. What would be the better way of showing the default value?
And it is not placeholder, it is text in a disabled input. If you click the checkbox, it enables the input and you can edit this value.

**Comment by marusak** [comment]
> There are also additional problems with the form using very wide entries for numbers. The input should be the approximate length of the expected value (plus a little bit of padding, to ensure numbers are not truncated). This should also be considered when the form is being worked.

Sure, I would day that 8 digits can be safely assumed for both as absolute maximum. (I would say that 5 would do just fine, but to be on a safe side)

**Comment by marusak** [comment]
@garrett bump, see my second to last comment.

**Comment by garrett** [comment]
Firstly, the dialog has changed since the screenshot. Here's a new one, which is even more problematic:

![Screenshot_2021-04-28 Podman containers - garrett Rain](https://user-images.githubusercontent.com/10246/116858322-7bae5e80-abfe-11eb-9ca2-88b7b9819d31.png)

(It has some other additional problems, with alignment of the memory limit and "with terminal" being an odd string placed and it's now in the memory limit section for some reason.)

> It shows what is the default value. What would be the better way of showing the default value?

Where does it show the default value? It doesn't do that anywhere in the screenshot.

>  And it is not placeholder, it is text in a disabled input. If you click the checkbox, it enables the input and you can edit this value.

Huh? No, the text in a disabled input is text in a disabled input, and that's fine. That has nothing to do with the placeholder values used for ports, volumes, and the environment settings, which is what this issue is about.

IP, host port, container port, host path, container path, key, and value are all supposed to be labels, but they're placeholder text instead, and that's wrong for several reasons, which are mentioned in the PF guidelines all over the place. I'll include some links below with some excerpts.

---

PF guidelines:

From https://www.patternfly.org/v4/components/form/design-guidelines/#placeholder-text-0:

> ### Placeholder Text
> <img src="https://www.patternfly.org/v4/images/placeholder-text.3788d4706023abd3baf00f4682e83230.png" alt="Example of a date placeholder text inside text input field" class="ws-img" width="323" height="104">
> 
> Within a form field, placeholder text provides an example of the required or recommended format for text input.
> 
> Use placeholder text when:
> 
> - The requested data type requires data to be inputted with specific format or syntax requirements.
> - The requested data type might be unfamiliar to the user.
> 
> Never use placeholder text to replace field labels or communicate crucial information since it disappears as soon as a user begins typing in a field.
> 
> If your placeholder text shares an example, precede it with “Example,” followed by a space so that users accessing the form with screen readers will know the difference between the field label and the provided placeholder text.
> 
> Examples:
> 
> - *Example, John Smith*
> - *Example, (555) 555-5555*
> - *MM-DD-YYYY*
> 
> <img src="https://www.patternfly.org/v4/images/placeholder-example.986dbdb26e17b4cfcfcefdeb8d284c68.png" alt="Example of a phone number placeholder text inside text input field" class="ws-img" width="323" height="120"><img src="https://www.patternfly.org/v4/images/placeholder-format.042431b3ac1dfdbb4edf2489fa9bb99e.png" alt="Example of a date placeholder text inside text input field" class="ws-img" width="323" height="120">

It's reiterated in https://www.patternfly.org/v4/components/form/design-guidelines/#placeholder-text-0 (but it's mainly the same, slightly rephrased and restructured).

And it's under accessibility too; from https://www.patternfly.org/v4/components/form/design-guidelines/#placeholder-text-1:

> Never write essential information in placeholder text. Screen readers like JAWS and NVDA don’t read it, nor is it widely supported among other assistive technologies. Additionally, placeholder text often shows in a low gray-on-white contrast, which may be inaccessible for users with visual impairments. Treat placeholder text as a visual supplement –– any crucial formatting requirements should be written in helper text so that users with assistive technology won’t miss it.

---

So, the most relevant points for this dialog are:
- Placeholder should just have examples or formatting guides, not what the entry should contain (which is a label's purpose)
- We're using placeholder as labels; this breaks a11y as stated above (for screen reader users, people with all sorts of vision problems (including common ones like astigmatism), and cognitive issues (such as memory impairment)).

---

FWIW: I was working on the redesign of this dialog on Thursday. I think it already has some scaling issues, with so much crammed in one dialog, especially if ports, volumes, and environment grow more than one line each... so I'm looking into ways to solve this, at least with a redesign.

**Comment by jelly** [comment]
Placeholders are now gone from the integrations tab:

![image](https://user-images.githubusercontent.com/67428/140277210-17f0ad89-feb6-4d01-9736-cd44dc7d1877.png)


Only in the Details tab we have a place holder for Command, we should get rid of it and that should resolve this issue right?

![image](https://user-images.githubusercontent.com/67428/140277403-b4faf99a-ea1b-465b-ab84-41b09291987e.png)

**Comment by garrett** [comment]
> Only in the Details tab we have a place holder for Command, we should get rid of it and that should resolve this issue right?

Well, the placeholder for command would be the default command or an example of a command. (Not the label.) https://www.patternfly.org/v4/components/form/design-guidelines/#placeholder-text-0

Or we could just remove it.

Meanwhile, host path in the integrations tab has placeholders.

And the alignment is super weird and wrong in the integrations tab too (but it's not placeholder related; it's for another issue).

**Comment by jelly** [comment]
> > Only in the Details tab we have a place holder for Command, we should get rid of it and that should resolve this issue right?
> 
> Well, the placeholder for command would be the default command or an example of a command. (Not the label.) https://www.patternfly.org/v4/components/form/design-guidelines/#placeholder-text-0
> Or we could just remove it.

Let me check if podman has a default, I think that depends on the container image which makes it all a bit fuzzy so probably better to leave out. 

> Meanwhile, host path in the integrations tab has placeholders.

That placeholder comes from cockpit actually the `FileAutoComplete` which we can't override as if it's empty it will default to what is shown.

> And the alignment is super weird and wrong in the integrations tab too (but it's not placeholder related; it's for another issue).

Yes, I noticed that too, can be improved for sure!

**Comment by marusak** [comment]
This has been fixed in the latest redesign:

![Screenshot from 2021-11-25 12-00-34](https://user-images.githubusercontent.com/12330670/143429630-626d762c-1b51-4499-9471-ad745fcb826d.png)

**Comment by garrett** [comment]
Not related to the issue, but related to the most recent screenshot:

Why is the host path missing an outline?

**Comment by jelly** [comment]
> Not related to the issue, but related to the most recent screenshot:
> 
> Why is the host path missing an outline?

It seems to be a fluke or a focus issue, I can't reproduce it on `main`.
